### PR TITLE
improve error handling and logging

### PR DESCRIPTION
**Pull request by ArielLima**
Improve error handling and logging, covering resources that fail to be deleted.

Card -> https://issues.redhat.com/browse/OSD-4827

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -26,26 +30,37 @@ const (
 
 var (
 	supportedRegions = []string{"us-east-1", "us-east-2", "us-west-1", "us-west-2", "ca-central-1", "eu-central-1", "eu-west-1", "eu-west-2", "eu-west-3", "ap-northeast-1", "ap-northeast-2", "ap-south-1", "ap-southeast-1", "ap-southeast-2", "sa-east-1"}
+	log              = logf.Log.WithName("shredder_logger")
 )
 
 func main() {
+	// Add the zap logger flag set to the CLI. The flag set must
+	// be added before calling pflag.Parse().
+	pflag.CommandLine.AddFlagSet(zap.FlagSet())
```
Comment (fahlmant) [inline main.go:39]: Why is this using flags? This isn't a cli tool

File: main.go [hunk 2]
```diff
@@ -57,38 +62,36 @@ func main() {
...
 		if err != nil {
-			fmt.Println("ERROR: ", err)
+			log.Error(err, "Failed to retieve list of accounts to clean")
 		}
 
 		for _, account := range accountCRList {
-			fmt.Println("Now Processing :")
-			fmt.Println("AccountCR:", account.Name)
-			fmt.Println("AccountID:", account.Spec.AwsAccountID)
+			logger := log.WithValues("AccountName", account.Name, "AccountID", account.Spec.AwsAccountID)
```
Comment (fahlmant) [inline main.go:88]: I think this message needs a bit more context as to what's happening

File: main.go [hunk 3]
```diff
@@ -57,38 +62,36 @@ func main() {
...
 
 			// assuming roles for the given AccountID
 			RoleArnParameter := "arn:aws:iam::" + account.Spec.AwsAccountID + ":role/OrganizationAccountAccessRole"
 			assumedRole, err := awsClient.AssumeRole(&sts.AssumeRoleInput{RoleArn: aws.String(RoleArnParameter), RoleSessionName: aws.String(sessionName)})
 			if err != nil {
-				fmt.Println("ERROR:", err)
+				logger.Error(err, "Failed to assume necessary account role")
```
Comment (fahlmant) [inline main.go:94]: Can you add the role name here?

File: main.go [hunk 4]
```diff
@@ -99,20 +102,20 @@ func main() {
 			assumedSessionToken := *assumedRole.Credentials.SessionToken
 
 			for _, region := range supportedRegions {
-				fmt.Println("\n Current Region : ", region)
+				logger = log.WithValues("AccountName", account.Name, "AccountID", account.Spec.AwsAccountID, "Region", region)
```
Comment (fahlmant) [inline main.go:105]: I think there needs to be more context here

File: main.go [hunk 5]
```diff
@@ -99,20 +102,20 @@ func main() {
...
 				assumedRoleClient, err := clientpkg.NewClient(assumedAccessKey, assumedSecretKey, assumedSessionToken, region)
 				if err != nil {
-					fmt.Println("ERROR:", err)
+					logger.Error(err, "Failed to initialize new client")
```
Comment (fahlmant) [inline main.go:108]: new client for what?

File: pkg/awsManager/ebsManager.go [hunk 6]
```diff
@@ -40,40 +41,39 @@ func ListEbsSnapshotForDeletion(client clientpkg.Client) []*string {
...
 	if ebsSnapshotsNotDeleted != nil {
-		return errors.New("Could not delete all snapshots for this region")
+		return errors.New("FailedComprehensiveSnapshotDeletion")
 	}
 
 	return nil
 }
 
-func ListVolumeForDeletion(client clientpkg.Client) []*string {
+func listVolumeForDeletion(client clientpkg.Client, logger logr.Logger) []*string {
```
Comment (fahlmant) [inline pkg/awsManager/ebsManager.go:68]: Why was this changed to unexported?

File: pkg/awsManager/ebsManager.go [hunk 7]
```diff
@@ -93,48 +93,48 @@ func ListVolumeForDeletion(client clientpkg.Client) []*string {
 	return ebsVolumesToBeDeleted
 }
 
-func DeleteEbsVolumes(client clientpkg.Client, ebsVolumesToBeDeleted []*string) error {
+func deleteEbsVolumes(client clientpkg.Client, ebsVolumesToBeDeleted []*string, logger logr.Logger) error {
```
Comment (fahlmant) [inline pkg/awsManager/ebsManager.go:96]: Why was this changed to unexported?

File: pkg/awsManager/ebsManager.go [hunk 8]
```diff
@@ -40,40 +41,39 @@ func ListEbsSnapshotForDeletion(client clientpkg.Client) []*string {
...
+func listVolumeForDeletion(client clientpkg.Client, logger logr.Logger) []*string {
 
 	var token *string
 	var ebsVolumesToBeDeleted []*string
 
 	for {
 		ebsVolumeList, err := client.DescribeVolumes(&ec2.DescribeVolumesInput{NextToken: token})
 		if err != nil {
-			fmt.Print(err)
+			logger.Error(err, "Failed to retrieve Volume list")
```
Comment (fahlmant) [inline pkg/awsManager/ebsManager.go:76]: Maybe, "Failed to list volumes" ?

File: pkg/awsManager/ec2Manager.go [hunk 9]
```diff
@@ -2,25 +2,25 @@ package awsManager
...
+func ListEc2InstancesForDeletion(client clientpkg.Client, logger logr.Logger) []*string {
 
 	var EC2InstancesToBeDeleted []*string
 	token := ""
 	for {
 		ec2Descriptions, err := client.DescribeInstances(&ec2.DescribeInstancesInput{NextToken: aws.String(token)})
 		if err != nil {
-			fmt.Println("ERROR:", err)
-
+			logger.Error(err, "Failed to retrieve ec2 descriptions")
```
Comment (fahlmant) [inline pkg/awsManager/ec2Manager.go:23]: "Failed to list EC2 instances"

File: pkg/awsManager/ec2Manager.go [hunk 10]
```diff
@@ -29,18 +29,14 @@ func ListEc2InstancesForDeletion(client clientpkg.Client) []*string {
...
 						EC2InstancesToBeDeleted = append(EC2InstancesToBeDeleted, instance.InstanceId)
 						break
 					}
 
 					if (*tag.Key == "clusterAccountName" || *tag.Key == "clusterClaimLink" || *tag.Key == "clusterNamespace" || *tag.Key == "clusterClaimLinkNamespace") && (*instance.State.Code != 48) {
 						EC2InstancesToBeDeleted = append(EC2InstancesToBeDeleted, instance.InstanceId)
-
 						break
 					}
-					// no need to delete the instances if the above two 'if' conditions are not true
```
Comment (fahlmant) [inline pkg/awsManager/ec2Manager.go:42]: Any reason to remove the comment?

File: pkg/awsManager/ec2Manager.go [hunk 11]
```diff
@@ -2,25 +2,25 @@ package awsManager
...
+	"github.com/go-logr/logr"
 	clientpkg "github.com/openshift/aws-account-shredder/pkg/aws"
-	"strings"
 )
 
-// this lists all the instances that are eligible for deletion based on the tags and stored them in instances to be deleted
+// ListEc2InstancesForDeletion this lists all the instances that are eligible for deletion based on the tags and stored them in instances to be deleted
 // this only creates an array of pointers and does not delete the instances
-func ListEc2InstancesForDeletion(client clientpkg.Client) []*string {
+func ListEc2InstancesForDeletion(client clientpkg.Client, logger logr.Logger) []*string {
```
Comment (fahlmant) [inline pkg/awsManager/ec2Manager.go:16]: Why was this unexported?

File: pkg/awsManager/vpcVpnManager.go [hunk 12]
```diff
@@ -35,111 +37,124 @@ func ListVPCforDeletion(client clientpkg.Client) ([]*string, error) {
...
 	}
-	fmt.Println("VPC's deleted successfully for this region  ")
+	logger.Info("VPC's deleted successfully")
 	return nil
-
 }
 
-func DeleteELB(client clientpkg.Client, vpcID *string) error {
-
+func deleteELB(client clientpkg.Client, vpcID *string, logger logr.Logger) error {
```
Comment (fahlmant) [inline pkg/awsManager/vpcVpnManager.go:157]: why was this unexported?

File: pkg/awsManager/vpcVpnManager.go [hunk 13]
```diff
@@ -166,13 +179,10 @@ func DeleteELB(client clientpkg.Client, vpcID *string) error {
...
 
 	}
-
 	return nil
-
 }
 
-func DeleteNatgateway(client clientpkg.Client, vpcID *string) error {
-
+func deleteNatgateway(client clientpkg.Client, vpcID *string, logger logr.Logger) error {
```
Comment (fahlmant) [inline pkg/awsManager/vpcVpnManager.go:185]: why was this unexported?

File: pkg/awsManager/vpcVpnManager.go [hunk 14]
```diff
@@ -199,13 +206,10 @@ func DeleteNatgateway(client clientpkg.Client, vpcID *string) error {
...
 		}
 	}
-
 	return nil
-
 }
 
-func DeleteNetworkLoadBalancer(client clientpkg.Client, vpcID *string) error {
-
+func deleteNetworkLoadBalancer(client clientpkg.Client, vpcID *string, logger logr.Logger) error {
```
Comment (fahlmant) [inline pkg/awsManager/vpcVpnManager.go:212]: why was this unexported?

File: pkg/awsManager/vpcVpnManager.go [hunk 15]
```diff
@@ -231,19 +233,16 @@ func DeleteNetworkLoadBalancer(client clientpkg.Client, vpcID *string) error {
 			break
 		}
 	}
-
 	return nil
 }
 
-func DetachAndDeleteNetworkInterface(client clientpkg.Client, vpcID *string) error {
+func detachAndDeleteNetworkInterface(client clientpkg.Client, vpcID *string, logger logr.Logger) error {
```
Comment (fahlmant) [inline pkg/awsManager/vpcVpnManager.go:239]: why was this unexported?

File: pkg/awsManager/vpcVpnManager.go [hunk 16]
```diff
@@ -274,19 +269,16 @@ func DetachAndDeleteNetworkInterface(client clientpkg.Client, vpcID *string) err
 			break
 		}
 	}
-
 	return nil
 }
 
-func DeleteGateway(client clientpkg.Client, vpcID *string) error {
-
+func deleteGateway(client clientpkg.Client, vpcID *string, logger logr.Logger) error {
```
Comment (fahlmant) [inline pkg/awsManager/vpcVpnManager.go:275]: why was this unexported?

File: pkg/awsManager/vpcVpnManager.go [hunk 17]
```diff
@@ -317,29 +305,24 @@ func DeleteGateway(client clientpkg.Client, vpcID *string) error {
 			break
 		}
 	}
-
 	return nil
 }
 
-func DeleteSubnetsForVPC(client clientpkg.Client, vpcId *string) error {
-
+func deleteSubnetsForVPC(client clientpkg.Client, vpcId *string, logger logr.Logger) error {
```
Comment (fahlmant) [inline pkg/awsManager/vpcVpnManager.go:311]: why was this unexported?

File: pkg/awsManager/vpcVpnManager.go [hunk 18]
```diff
@@ -350,19 +333,16 @@ func DeleteSubnetsForVPC(client clientpkg.Client, vpcId *string) error {
 			break
 		}
 	}
-
 	return nil
 }
 
-func DeleteRouteTables(client clientpkg.Client, vpcId *string) error {
-
+func deleteRouteTables(client clientpkg.Client, vpcId *string, logger logr.Logger) error {
```
Comment (fahlmant) [inline pkg/awsManager/vpcVpnManager.go:339]: why was this unexported?

File: pkg/awsManager/vpcVpnManager.go [hunk 19]
```diff
@@ -401,25 +378,22 @@ func DeleteRouteTables(client clientpkg.Client, vpcId *string) error {
 	return nil
 }
 
-func DeleteNetworkAcl(client clientpkg.Client, vpcId *string) error {
+func deleteNetworkAcl(client clientpkg.Client, vpcId *string, logger logr.Logger) error {
```
Comment (fahlmant) [inline pkg/awsManager/vpcVpnManager.go:381]: why was this unexported?

File: pkg/awsManager/vpcVpnManager.go [hunk 20]
```diff
@@ -434,24 +408,22 @@ func DeleteNetworkAcl(client clientpkg.Client, vpcId *string) error {
 	return nil
 }
 
-func DeleteSecurityGroups(client clientpkg.Client, vpcId *string) error {
+func deleteSecurityGroups(client clientpkg.Client, vpcId *string, logger logr.Logger) error {
```
Comment (fahlmant) [inline pkg/awsManager/vpcVpnManager.go:411]: why was this unexported?

File: pkg/awsManager/vpcVpnManager.go [hunk 21]
```diff
@@ -508,60 +477,49 @@ func DeleteVpcEndpoint(client clientpkg.Client, vpcId *string) error {
...
 		}
-
 	}
 
 	return nil
 
 }
 
-func DetachVpnGateway(client clientpkg.Client, vpcId *string) error {
+func detachVpnGateway(client clientpkg.Client, vpcId *string, logger logr.Logger) error {
```
Comment (fahlmant) [inline pkg/awsManager/vpcVpnManager.go:509]: why was this unexported?

File: main.go (hunk 2 above)
Comment (ArielLima) [reply main.go:88]: That is not a message, I simply made the logger always display the accountCR and accountID so there is no longer a need to output that information beforehand.

File: main.go (hunk 3 above)
Comment (ArielLima) [reply main.go:94]: +1

File: main.go (hunk 2 above)
Comment (fahlmant) [reply main.go:88]: Okay. Then please add a message at the start of processing each account CR

Comment (fahlmant) [reply main.go:88]: Okay. I still think its helpful to have a log that says "Starting cleanup of Account X"

File: main.go (hunk 4 above)
Comment (ArielLima) [reply main.go:105]: All this line is doing is updating the logger to always display the region when logging. No longer a need to display the region being processed beforehand

Comment (fahlmant) [reply main.go:105]: I think its still helpful to say "Cleaning region y"

File: main.go (hunk 5 above)
Comment (ArielLima) [reply main.go:108]: +1

File: pkg/awsManager/ebsManager.go (hunk 6 above)
Comment (ArielLima) [reply pkg/awsManager/ebsManager.go:68]: because it is a helper function of `CleanEbsVolumes()`

File: pkg/awsManager/ebsManager.go (hunk 7 above)
Comment (ArielLima) [reply pkg/awsManager/ebsManager.go:96]: because it is a helper function of CleanEbsVolumes()

File: pkg/awsManager/ebsManager.go (hunk 8 above)
Comment (ArielLima) [reply pkg/awsManager/ebsManager.go:76]: I think "retrieve volume list" is more true to what is happening but am willing to change it if you think it better.

Comment (fahlmant) [reply pkg/awsManager/ebsManager.go:76]: Just trying to keep all the errors consistent

File: pkg/awsManager/ebsManager.go (hunk 6 above)
Comment (fahlmant) [reply pkg/awsManager/ebsManager.go:68]: It has uses cases outside of CleanEbsVolumes(), and all the other AWS resource types also had them exported (some still do too)

File: pkg/awsManager/ec2Manager.go (hunk 9 above)
Comment (ArielLima) [reply pkg/awsManager/ec2Manager.go:23]: +1

File: pkg/awsManager/ec2Manager.go (hunk 10 above)
Comment (ArielLima) [reply pkg/awsManager/ec2Manager.go:42]: I don't think writing comments for previous lines should be acceptable, I will add a comment above the statements explaining what is happening. Also this method needs to be unexported because it is just a helper to CleanEc2Instances()

File: pkg/awsManager/ec2Manager.go (hunk 11 above)
Comment (ArielLima) [reply pkg/awsManager/ec2Manager.go:16]: It wasn't, just added logger as a parameter

Comment (ArielLima) [reply pkg/awsManager/ec2Manager.go:16]: probably should be becuase it isn't and shouldn't be used anywhere outside this file

File: pkg/awsManager/vpcVpnManager.go (hunk 12 above)
Comment (ArielLima) [reply pkg/awsManager/vpcVpnManager.go:157]: because it is a helper method that isn't and shouldn't be used outside this file

File: pkg/awsManager/vpcVpnManager.go (hunk 13 above)
Comment (ArielLima) [reply pkg/awsManager/vpcVpnManager.go:185]: because it is a helper method that isn't and shouldn't be used outside this file

File: pkg/awsManager/vpcVpnManager.go (hunk 14 above)
Comment (ArielLima) [reply pkg/awsManager/vpcVpnManager.go:212]: because it is a helper method that isn't and shouldn't be used outside this file

File: pkg/awsManager/vpcVpnManager.go (hunk 15 above)
Comment (ArielLima) [reply pkg/awsManager/vpcVpnManager.go:239]: because it is a helper method that isn't and shouldn't be used outside this file

File: pkg/awsManager/vpcVpnManager.go (hunk 16 above)
Comment (ArielLima) [reply pkg/awsManager/vpcVpnManager.go:275]: because it is a helper method that isn't and shouldn't be used outside this file

File: pkg/awsManager/vpcVpnManager.go (hunk 17 above)
Comment (ArielLima) [reply pkg/awsManager/vpcVpnManager.go:311]: because it is a helper method that isn't and shouldn't be used outside this file

File: pkg/awsManager/vpcVpnManager.go (hunk 18 above)
Comment (ArielLima) [reply pkg/awsManager/vpcVpnManager.go:339]: because it is a helper method that isn't and shouldn't be used outside this file

File: pkg/awsManager/vpcVpnManager.go (hunk 19 above)
Comment (ArielLima) [reply pkg/awsManager/vpcVpnManager.go:381]: because it is a helper method that isn't and shouldn't be used outside this file

File: pkg/awsManager/vpcVpnManager.go (hunk 20 above)
Comment (ArielLima) [reply pkg/awsManager/vpcVpnManager.go:411]: because it is a helper method that isn't and shouldn't be used outside this file

File: pkg/awsManager/vpcVpnManager.go (hunk 21 above)
Comment (ArielLima) [reply pkg/awsManager/vpcVpnManager.go:509]: because it is a helper method that isn't and shouldn't be used outside this file

File: main.go (hunk 2 above)
Comment (ArielLima) [reply main.go:88]: I disagree. All that does is create a log that we will never look at and remove later since the information we want will always accessible in other useful logs.

File: pkg/awsManager/ebsManager.go (hunk 6 above)
Comment (ArielLima) [reply pkg/awsManager/ebsManager.go:68]: Maybe I am misunderstanding something. The methods I unexported were all helper methods not used anywhere else and that shouldn't be used for anything other than their `Clean` parent function

File: main.go (hunk 2 above)
Comment (fahlmant) [reply main.go:88]: This makes it easy to see when a run started on an Account CR. I can just grep for "starting cleanup of osd-mgmt-aaabbb" and then go from there. Without this line, its hard to parse where one run ends and another begins.

**Comment by fahlmant** [comment]
@ArielLima  Why shouldn't those functions be used outside of their files? They're all fairly helpful on their own and I don't see a good enough justification to unexport them since they're already exported.

File: pkg/awsManager/ebsManager.go (hunk 6 above)
Comment (fahlmant) [reply pkg/awsManager/ebsManager.go:68]: But why shouldn't they ever? I know they aren't at the moment but I don't think that's a good reason to say they shouldn't

File: main.go (hunk 2 above)
Comment (sam-nguyen7) [reply main.go:88]: I'm +1 for having more context here. For me personally it would be easier to search/find the starting point vs looking through all the logs with the accountcr and id. What's the issue with having both the message and the logger? I don't think it would be an overkill @ArielLima

Comment (ArielLima) [reply main.go:88]: I still don't think it will be that useful but I'll add it

File: pkg/awsManager/ec2Manager.go (hunk 10 above)
Comment (fahlmant) [reply pkg/awsManager/ec2Manager.go:42]: Please add a comment in its place then. It's explaining the code, I'm not sure the position really matters if its before or after. Regardless, I think this explanation is good to have somewhere

**Comment by ArielLima** [comment]
@fahlmant The way things are right now we have different managers for different AWS resources we want to clean. Each manager has a `Clean` method that will list + delete those resources. I don't see why we would want to export the helpers `list` and `delete` unless we have something in mind for them now since we can just use `Clean`.

**Comment by ArielLima** [comment]
I can re-export them since you believe it's important but I don't think we should be exporting something just because it seems nice. We know it works with clean but some of those List methods list only very specific things we want the shredder to be able to delete.

**Comment by fahlmant** [comment]
I still don't understand why that means we NEED to unexport them. It doesn't make the code cleaner, just different. The change doesn't actually gain anything

File: main.go [hunk 22]
```diff
@@ -86,12 +86,12 @@ func main() {
 
 		for _, account := range accountCRList {
 			logger := log.WithValues("AccountName", account.Name, "AccountID", account.Spec.AwsAccountID)
-
+			logger.Info("New Account being shredder") // Usefull for keeping track of when work begins on an account
```
Comment (fahlmant) [inline main.go:89]: Shredded

File: main.go [hunk 23]
```diff
@@ -105,7 +105,7 @@ func main() {
 				logger = log.WithValues("AccountName", account.Name, "AccountID", account.Spec.AwsAccountID, "Region", region)
 				assumedRoleClient, err := clientpkg.NewClient(assumedAccessKey, assumedSecretKey, assumedSessionToken, region)
```
Comment (fahlmant) [inline main.go:106]: Please add a log before this stating "Beginning cleanup of region $REGION"

File: main.go (hunk 22 above)
Comment (fahlmant) [reply main.go:89]: Probably should read something like "Account $ACCOUNT_ID from $ACCOUNT_CR being shredded"

Comment (ArielLima) [reply main.go:89]: Thanks for pointing out the typo. 
I am sorry but it makes no sense to add account_id and account_cr in this log since those two pieces of information are already being logged `logger := log.WithValues("AccountName", account.Name, "AccountID", account.Spec.AwsAccountID)`

Comment (ArielLima) [reply main.go:89]: I thought you wanted this so you could easily see the beginning of an account shred but if you already have the account id and cr name than this is useless

Comment (fahlmant) [reply main.go:89]: Say I'm grepping through logs. 
`oc logs podname | grep "being shredded" | grep account_cr_name`
vs
`oc logs podname| grep "account_cr_name being shredded"`

Simplifies searching to a single string than having to grep for multiple strings. The same thought can be applied to a jq select.

**Review by fahlmant — APPROVED** [review]
/hold cancel
/lgtm

File: main.go (hunk 22 above)
Comment (ArielLima) [reply main.go:89]: `oc logs podname | grep account_cr_name`

Comment (fahlmant) [reply main.go:89]: But that doesn't automatically show you the top of a specific run. An account CR will be run multiple times in the operator logs
